File: hojoo/week1/long_jump.py
# 접근: dp
# 첫 번째 칸부터 순회하면 칸에 도달할 때 마다 해당 칸에 도달할 수 있는 방법의 수가 확정됨
# dp[i] = i칸에 도달할 수 있는 방법의 수
# 시간 복잡도: O(n)

def solution(n):
    dp = [0] * n
    
    # n이 1일땐 1반환
    if n == 1:
        return 1
    dp[0] = 1
    dp[1] = 1
    
    for i in range(n-1):
        dp[i+1] += dp[i]
        try:
            dp[i+2] += dp[i]
        except:
            continue
    
    return dp[-1] % 1234567


if __name__ == "__main__":
    print(solution(4))
    print(solution(3))
    print(solution(1))

File: jihwan/week1/LongJump.cpp
#include <string>
#include <vector>

using namespace std;

long long DP[2001]; //i번째 칸에 올 수 있는 경우의 수 저장

void dp(int n) {
    DP[1]=1; //초기값
    DP[2]=1;
    for(int i=1; i<n; i++) {
        if(i+1<=n) { // 1칸 이동하는 경우 현재칸까지 오는 경우의수를 i+1칸에 더해줌
            DP[i+1] = (DP[i+1]+DP[i])%1234567;
        }
        if(i+2<=n) { //위와 동일
            DP[i+2] = (DP[i+2]+DP[i])%1234567;
        }
    }
}
long long solution(int n) {
    long long answer = 0;
    dp(n);
    answer = DP[n];
    return answer;
}